### PR TITLE
feat(diff): add zombie & shadow API diff support

### PR DESCRIPTION
**Pull request by FrimIdan**
Related PR https://github.com/apiclarity/speculator/pull/12

### CODE REVIEW

File: backend/pkg/backend/backend.go [hunk 1]
```diff
@@ -306,13 +310,51 @@ func (b *Backend) handleHTTPTrace(trace *_spec.SCNTelemetry) error {
...
 	}
 
+	event.SpecDiffType = getSpecDiffType(providedDiffType, reconstructedDiffType)
+
 	_database.CreateAPIEvent(event)
 
 	return nil
 }
 
+func getAPIDiffType(diffType _spec.DiffType) models.DiffType {
```
Comment (fishkerez) [inline backend/pkg/backend/backend.go:323]: maybe convertAPIDiffType?

File: backend/pkg/backend/backend.go [hunk 2]
```diff
@@ -306,13 +310,51 @@ func (b *Backend) handleHTTPTrace(trace *_spec.SCNTelemetry) error {
...
+var diffTypePriority = map[models.DiffType]int{
+	// starting from 1 since unknown type will return 0
+	models.DiffTypeNODIFF:      1,
+	models.DiffTypeGENERALDIFF: 2,
+	models.DiffTypeSHADOWDIFF:  3,
+	models.DiffTypeZOMBIEDIFF:  4,
+}
+
+// getSpecDiffType will return the type with the highest priority.
+func getSpecDiffType(providedDiff, reconstructedDiff models.DiffType) models.DiffType {
```
Comment (fishkerez) [inline backend/pkg/backend/backend.go:350]: maybe getHighestPrioritySpecDiffType?

Comment (fishkerez) [inline backend/pkg/backend/backend.go:350]: providedDiff, reconstructedDiff => providedDiffType, reconstructedDiffType

File: backend/pkg/backend/backend.go [hunk 3]
```diff
@@ -306,13 +310,51 @@ func (b *Backend) handleHTTPTrace(trace *_spec.SCNTelemetry) error {
 			event.NewProvidedSpec = string(modified)
 		}
 		event.PathID = providedDiff.PathID
+		providedDiffType = getAPIDiffType(providedDiff.Type)
 	}
 
+	event.SpecDiffType = getSpecDiffType(providedDiffType, reconstructedDiffType)
```
Comment (fishkerez) [inline backend/pkg/backend/backend.go:316]: Why are we taking the higher specDiffType, and not creating providedspecDiffType and reconstructedspecDiffType on the ApiEvent?

Comment (FrimIdan) [reply backend/pkg/backend/backend.go:316]: yes by design

File: backend/pkg/backend/backend.go (hunk 2 above)
Comment (akpsgit) [reply backend/pkg/backend/backend.go:350]: or getSeverestDiffType

File: backend/pkg/backend/backend.go (hunk 3 above)
Comment (akpsgit) [reply backend/pkg/backend/backend.go:316]: @fishkerez I think that the DB query will be harder to merge the type on fetch.

**Review by akpsgit — APPROVED** [review]

**Pull request merged**